bin/create_clusters.py

```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/python
+
+import subprocess
+import json
+import shlex
+
+statestore = 'moj-cp-k8s-investigation-kops'
+domain = '.k8s.integration.dsd.io'
+
+
+def run(command):
+    process = subprocess.Popen(shlex.split(command), stdout=subprocess.PIPE)
+    while True:
+        output = process.stdout.readline()
+        if output == '' and process.poll() is not None:
+            break
+        if output:
+            print output.strip()
+    rc = process.poll()
+    return rc
+
+
+# terraform init
+try:
+    print(subprocess.check_output(['terraform', 'init']))
+except subprocess.CalledProcessError as e:
+    print('error initing terraform, exit.')
+
+# get terraform workspace list
+try:
+    workspaces = subprocess.check_output(
+        ['terraform', 'workspace', 'list']).decode('utf-8').split()
+    workspaces.remove('*')
+except subprocess.CalledProcessError as e:
+    print('error listing workspaces, exit.')
+
+try:
+    yaml = [s.strip('.yaml') for s in subprocess.check_output(
+        ['ls', '../../kops/']).decode('utf-8').split()]
+except subprocess.CalledProcessError as e:
+    print('error listing cluster yaml files, exit.')
+
+# query the kops state store to list the active clusters
+try:
+    clusters = json.loads(str(subprocess.check_output(
+        ['kops', 'get', 'clusters', '--state=s3://' + statestore,
+         '--output=json']).decode('utf-8')))
+except subprocess.CalledProcessError as e:
+    print('error listing clusters, exit.')
+
+# generate a list of cluster names only
+clusterlist = []
+for item in clusters:
+    clusterlist.append(item['metadata']['name'].replace(domain, ''))
+
+# create the corresponding clusters
+for item in yaml:
+    if item in workspaces and item in clusterlist:
+        print('cluster ' + item + ' already exists, skipping')
+    else:
+        print('creating cluster ' + item)
+        run('../../bin/create_cluster.sh ' + item)
+
+print('Done.')
```

Comment:
After a very brief chat with @kalbir just now I think this probably warrants some discussion - in terms of how we handle the config of multiple clusters, how/when those configs are applied, and what testing or previewing config changes looks like. We could have individual `cluster.yml` files for each individual cluster, but then that raises the question of what should be in there and what should be passed in at YML render time - ideally we want to be able to test cluster config changes without needing to apply them to a specific running cluster, so we'd need to be careful to keep things like cluster and DNS names out of it.

The other question is if we should have a single script or process that applies all changes to all clusters in one go - that doesn't feel right to me as it'd require a one-to-one mapping between `cluster.yml` files and deployed clusters, preventing us from previewing or testing changes, but I'm open to suggestions here.

Reply:
Kerin, if you have a pipeline that creates/updates a specific cluster, you cannot run that for multiple clusters, because its designed to work with a specific cluster.
You can have however a generic pipeline that takes as an argument the cluster name and creates/updates that given cluster, and then you can reuse that pipeline for different clusters.
we dont want to reuse the same pipeline code over and over because that will generate code duplication.

the one to one mapping between cluster.yaml and the deployed clusters is what we already have on this repo (sandbox cluster and non-prod cluster), inside kops, so that is why is used that approach for this feature.

again, we can have a generic pipeline manually triggered from witch you specify as argument the cluster name, but the issue with that is that you cant guarantee the cluster.yaml in the repo will match the running cluster config. The running cluster config may be outdated if the pipeline did not run for that given cluster after a commit was pushed to master.
with this approach everything that is in the repo is applied on the cluster and you can consider the repo the source of truth.
You can have a 'testing' branch of this repo to apply it against a testing subset of clusters to test the config before applying it to the production clusters.